server2/Arcadia.Assistant/Arcadia.Assistant.Sharepoint/SharepointSynchronizator.cs

```diff
@@ -0,0 +1,170 @@
+﻿namespace Arcadia.Assistant.Sharepoint
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using CalendarEvent;
+
+    using Employees.Contracts;
+
+    using ExternalStorages.Abstractions;
+
+    using Microsoft.Extensions.Logging;
+
+    using SickLeaves.Contracts;
+
+    using Vacations.Contracts;
+
+    using WorkHoursCredit.Contracts;
+
+    public sealed class SharepointSynchronizator
+    {
+        #region ctor
+
+        public SharepointSynchronizator(ISickLeaves sickLeaves, IVacations vacations, IWorkHoursCredit workouts, ISharepointDepartmentsCalendarsSettings departmentsCalendarsSettings, ILogger? logger)
+        {
+            this.sickLeaves = sickLeaves;
+            this.vacations = vacations;
+            this.workouts = workouts;
+            this.departmentsCalendarsSettings = departmentsCalendarsSettings;
+            this.logger = logger;
+        }
+
+        #endregion
+
+        #region public interface
+
+        public async Task Synchronize(IEmployees employees, IEnumerable<string> departments, IExternalStorage storage, CancellationToken cancellationToken)
+        {
+            var vacationsSync = new EmployeeVacationsSynchronization(storage, this.logger);
+            var workoutsSync = new EmployeeWorkoutsSynchronization(storage, this.logger);
+            var sickLeavesSync = new EmployeeSickLeavesSynchronization(storage, this.logger);
+
+            var sharepointItemsByCalendars = departments.Distinct().ToDictionary(x => x,
+                dId => this.GetSharepointCalendarsByDepartment(dId).ToDictionary(x => x,
+                    async cal => await this.GetAllSharepointItemsForCalendar(storage, cal)));
+
+            foreach (var departmentId in departments)
+            {
+                var departmentEmployees = await employees.FindEmployeesAsync(EmployeesQuery.Create().ForDepartment(departmentId), cancellationToken);
+                var employeeIds = departmentEmployees.Select(x => x.EmployeeId).ToArray();
+
+                var employeeVacationsTask = this.vacations.GetCalendarEventsByEmployeeAsync(employeeIds, cancellationToken);
+                var employeeWorkoutsTask = this.workouts.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                var employeeSickLeavesTask = this.sickLeaves.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                await Task.WhenAll(employeeVacationsTask, employeeWorkoutsTask, employeeSickLeavesTask);
+
+                var employeeVacationValues = employeeVacationsTask.Result.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.VacationId, CalendarEventTypes.Vacation), x => x);
```

Comment:
Could you, please, wrap these too long lines?

Reply:
Lines wrapped.